diff --git a/vendor/riscv/riscv-isa-sim/riscv/Proc.cc b/vendor/riscv/riscv-isa-sim/riscv/Proc.cc
index 7058a1b0..4795067f 100644
--- a/vendor/riscv/riscv-isa-sim/riscv/Proc.cc
+++ b/vendor/riscv/riscv-isa-sim/riscv/Proc.cc
@@ -170,13 +170,15 @@ st_rvfi Processor::step(size_t n, st_rvfi reference_) {
             }
 
             // If it is set or clear we need to inject also the value in the CSR
-            if ((rvfi.insn & MASK_CSRRC) == MATCH_CSRRC   ||
-                (rvfi.insn & MASK_CSRRS) == MATCH_CSRRS   ||
-                (rvfi.insn & MASK_CSRRCI) == MATCH_CSRRCI ||
-                (rvfi.insn & MASK_CSRRSI) == MATCH_CSRRSI) {
-                if (reference->csr_valid[INDEX_CSR(read_csr)]) {
-                    this->put_csr(read_csr, reference->csr_wdata[INDEX_CSR(read_csr)]);
-                    rvfi.csr_wdata[INDEX_CSR(read_csr)] = reference->csr_wdata[INDEX_CSR(read_csr)];
+            if (reg_t read_csr = this->get_state()->last_inst_fetched.rs1()) {
+                if ((rvfi.insn & MASK_CSRRC) == MATCH_CSRRC   ||
+                    (rvfi.insn & MASK_CSRRS) == MATCH_CSRRS   ||
+                    (rvfi.insn & MASK_CSRRCI) == MATCH_CSRRCI ||
+                    (rvfi.insn & MASK_CSRRSI) == MATCH_CSRRSI) {
+                    if (reference->csr_valid[INDEX_CSR(read_csr)]) {
+                        this->put_csr(read_csr, reference->csr_wdata[INDEX_CSR(read_csr)]);
+                        rvfi.csr_wdata[INDEX_CSR(read_csr)] = reference->csr_wdata[INDEX_CSR(read_csr)];
+                    }
                 }
             }
             break;
